### PR TITLE
mach: fix logic to override paths for legacy layout

### PR DESCRIPTION
**Pull request by mukilan**
wptrunner internally derives the path to the MANIFEST.json file from the `metadata_path` passed via `test_paths`. The current logic overrides only the `metadata_path`, leaving the manifest_path pointing at the wrong (non-legacy) layout's MANIFEST.json.

In #34436 we observed that the recent WPT imports create a transient `.cache` diretory that wptrunner logic uses for optimization. This has not been observed until #34436 because a [recent bump in MANIFEST.json's schema version][1] triggered the creation of the cache. Because of the above issue with incorrect path and the fact that we *first* trigger `mach update-wpt` for legacy layout during WPT import, the MANIFEST.json of non-legacy layout gets incorrectly migrated during the invocation of `update-wpt` for legacy layout but the cache is still created under legacy-layout's path as it is not based on `manifest_path`. The subsequent invocation of `mach update-wpt` for non-legacy finds the MANIFEST.json already migrated so the `.cache` directory is not constucted.

This change simply replaces the whole object using the wptrunner's `TestRoot` class constructor so that all derived paths are calculated correctly. We also add the `.cache` folders to gitignore as it seems like they are expected to be created during such version migrations.

[1]: https://github.com/web-platform-tests/wpt/pull/49406

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,6 +12,12 @@
 /python/_venv*
 /python/tidy/servo_tidy.egg-info
 /tests/wpt/sync
+/tests/wpt/meta/.cache
+/tests/wpt/meta-legacy-layout/.cache
+/tests/wpt/mozilla/meta/.cache
+/tests/wpt/mozilla/meta-legacy-layout/.cache
+/tests/wpt/webgl/meta/.cache
+/tests/wpt/webgl/meta-legacy-layout/.cache
```
Comment (sagudev) [inline .gitignore:20]: NIT: Maybe we can just ignore .cache folders.

**Review by sagudev — APPROVED** [review]
But let's test it with WPT-2013 first.

**Comment by sagudev** [comment]
I think something must be wrong as https://github.com/servo/servo/actions/runs/12147901179 has way less errors.

**Review by sagudev — CHANGES_REQUESTED** [review]
Just to prevent accidental landing.

**Comment by mukilan** [comment]
I was just looking into some of the errors. They seem to be due to the current meta-legacy-layout/MANIFEST.json containing tests that no longer exist (moved or deleted).

For example the `/WebCryptoAPI/derive_bits_keys/cfrg_curves_bits.https.any.html` was [moved](https://github.com/web-platform-tests/wpt/commit/203d2ac45900139633d4f3f29750ea4b5e06d6f5#diff-81c0678a498db0dd08be82576d1ffbeec6963b0c4a046c8a294519915d0dd179) a couple of months back, but legacy layout's manifest still has an entry for it:
```
   "derive_bits_keys": {
     "cfrg_curves_bits.https.any.js": [
      "c1837591ee85d40d7ccf98b6ef56285369f0d7fd",
      [
       "WebCryptoAPI/derive_bits_keys/cfrg_curves_bits.https.any.html",
       {
        "script_metadata": [
         [
          "title",
          "WebCryptoAPI: deriveBits() Using ECDH with CFRG Elliptic Curves"
         ],
         [
          "script",
          "cfrg_curves_bits_fixtures.js"
         ],
         [
          "script",
          "cfrg_curves_bits.js"
         ]
        ]
       }
      ],
 ```     

On `main` branch, without this fix, the non-legacy manifest is being used during run and since that is more up-to-date, we don't see failures for it.

I wonder if we were deliberately not overriding the MANIFEST.json path so that both engines share the same file. In that case, why do we have a separate MANIFEST.json file in `meta-legacy-layout`.

**Comment by mukilan** [comment]
I think the next WPT import should update the `meta-legacy-layout/MANIFEST.json` with the correct tests, so this change seems "safe" (I've [triggered an import job](https://github.com/servo/servo/actions/runs/12153469018) based on this branch to test), but it would be nice to confirm if the sharing of MANIFEST.json between engines was intentional or accidental e.g. we forgot to delete `meta-legacy-layout/MANIFEST.json` when we moved the folders to support the new engine. The last update to `meta-legacy-layout/MANIFEST.json` was [a year ago](https://github.com/servo/servo/commit/1ec308776c2352528ea680dcdd21b4be41086ef7), but it was during a WPT import process. I'm not sure why subsequent imports stopped updating the file - maybe the wptrunner code changed in a way that meant MANIFEST.json was being read from the non-overridden path.

@mrobinson do you know if the behavior of both engines sharing the `meta/MANIFEST.json` is  by design and we just forgot to delete the other MANIFEST.json?

**Comment by mukilan** [comment]
I checked out #34473 locally and validated that both legacy and non-legacy have the same information:

```
review on  wpt_update_04-12-2024 [$?] is 📦 v0.0.1 via 🦀 v1.80.1 
❯ diff tests/wpt/meta/MANIFEST.json tests/wpt/meta-legacy-layout/MANIFEST.json
review on  wpt_update_04-12-2024 [$?] is 📦 v0.0.1 via 🦀 v1.80.1 
❯ echo $?
0
```

**Comment by mukilan** [comment]
> Is there any reason that legacy layout's `MANIFEST.json` file would need to be different than the one for non-legacy layout

I don't see any reason for them to be different. But whether we should have 2 separate files or just 1 single file is not clear to me. Based on the wptrunner code it seems like having two separate files is the safe option - the current logic was sharing a single file "by accident" as it was mutating the (presumably private) internal state of the `TestRoot` objects. This PR uses avoids doing that but will result in two files, which doesn't seem bad as it also implies that the data for legacy and non-legacy are inside their separate directories. I can undo this change if we think we should stick with one file.

File: .gitignore (hunk 1 above)
Comment (mukilan) [reply .gitignore]: I was worried about it being too general and hiding stray files that we should be aware of. In [573213a](https://github.com/servo/servo/commit/573213a7559b31073e83c90f45850851084ff1bf) I've change this to use wildcards, but still scoped to `/tests/wpt/`.

**Comment by mukilan** [comment]
@mrobinson and I were discussing about the plans for removing WPT suppor for Layout 2013. I've created #34474 to track this. I think we can go ahead with this PR to unblock the WPT import job and work on #34474 separately.

**Review by sagudev — APPROVED** [review]

**Review by mrobinson — APPROVED** [review]

**Pull request merged**